src/golang/lib/workflow/engine/aq_engine.go

```diff
@@ -0,0 +1,811 @@
+import (
+	"context"
+	"fmt"
+	"path"
+	"reflect"
+	"time"
+
+	artifact_db "github.com/aqueducthq/aqueduct/lib/collections/artifact"
+	"github.com/aqueducthq/aqueduct/lib/collections/artifact_result"
+	"github.com/aqueducthq/aqueduct/lib/collections/notification"
+	operator_db "github.com/aqueducthq/aqueduct/lib/collections/operator"
+	"github.com/aqueducthq/aqueduct/lib/collections/operator_result"
+	"github.com/aqueducthq/aqueduct/lib/collections/shared"
+	"github.com/aqueducthq/aqueduct/lib/collections/user"
+	"github.com/aqueducthq/aqueduct/lib/collections/workflow"
+	"github.com/aqueducthq/aqueduct/lib/collections/workflow_dag"
+	"github.com/aqueducthq/aqueduct/lib/collections/workflow_dag_edge"
+	"github.com/aqueducthq/aqueduct/lib/collections/workflow_dag_result"
+	"github.com/aqueducthq/aqueduct/lib/collections/workflow_watcher"
+	"github.com/aqueducthq/aqueduct/lib/cronjob"
+	"github.com/aqueducthq/aqueduct/lib/database"
+	"github.com/aqueducthq/aqueduct/lib/job"
+	shared_utils "github.com/aqueducthq/aqueduct/lib/lib_utils"
+	"github.com/aqueducthq/aqueduct/lib/vault"
+	dag_utils "github.com/aqueducthq/aqueduct/lib/workflow/dag"
+	"github.com/aqueducthq/aqueduct/lib/workflow/operator"
+	"github.com/aqueducthq/aqueduct/lib/workflow/operator/connector/github"
+	workflow_utils "github.com/aqueducthq/aqueduct/lib/workflow/utils"
+	"github.com/dropbox/godropbox/errors"
+	"github.com/google/uuid"
+	log "github.com/sirupsen/logrus"
+)
+
+type AqueductTimeConfig struct {
+	// Configures exactly long we wait before polling again on an in-progress operator.
+	OperatorPollInterval time.Duration
+
+	// Configures the maximum amount of time we wait for execution to finish before aborting the run.
+	ExecTimeout time.Duration
+
+	// Configures the maximum amount of time we want for any leftover, in-progress operators to complete,
+	// after execution has already finished. Once this time is exceeded, we'll give up.
+	CleanupTimeout time.Duration
+}
+
+type EngineReaders struct {
+	WorkflowReader          workflow.Reader
+	WorkflowDagReader       workflow_dag.Reader
+	WorkflowDagEdgeReader   workflow_dag_edge.Reader
+	WorkflowDagResultReader workflow_dag_result.Reader
+	OperatorReader          operator_db.Reader
+	OperatorResultReader    operator_result.Reader
+	ArtifactReader          artifact_db.Reader
+	ArtifactResultReader    artifact_result.Reader
+	UserReader              user.Reader
+}
+
+type EngineWriters struct {
+	WorkflowWriter          workflow.Writer
+	WorkflowDagWriter       workflow_dag.Writer
+	WorkflowDagEdgeWriter   workflow_dag_edge.Writer
+	WorkflowDagResultWriter workflow_dag_result.Writer
+	WorkflowWatcherWriter   workflow_watcher.Writer
+	OperatorWriter          operator_db.Writer
+	OperatorResultWriter    operator_result.Writer
+	ArtifactWriter          artifact_db.Writer
+	ArtifactResultWriter    artifact_result.Writer
+	NotificationWriter      notification.Writer
+}
+
+type aqEngine struct {
+	database       database.Database
```

Comment:
I would consider using public fields. It's good enough to control access using top-level private type.  One past experience is #234 . There are two benefits both reflected in the mentioned PR:
* easier to dump the entire struct to json string for future logging and debugging. Json marshal will hide private fields.
* It's easier to move things around before codebase get stable. There's less renaming you would need to worry when one day you suddenly want to expose some field